awx/main/dispatch/worker/base.py

```diff
@@ -152,9 +157,18 @@ def run(self, *args, **kwargs):
                         self.process_task(json.loads(e.payload))
                     if self.should_stop:
                         return
+                    self.pg_is_down = False
             except psycopg2.InterfaceError:
                 logger.warn("Stale Postgres message bus connection, reconnecting")
                 continue
+            except (db.DatabaseError, psycopg2.OperationalError):
+                if not self.pg_is_down:
+                    logger.exception("Postgres error processing tasks")
+                    self.pg_down_time = time.time()
+                    self.pg_is_down = True
```

Comment:
After this, I believe it would be best to pause for a second or two, so that we're not pummeling a recovering database with failing requests. However, while this is happening we probably need to have a sub-loop checking `self.should_stop` at a faster rate.